File: code/cpp/src/parser.cpp
#include "tlsfp/parser.hpp"
#include <cstring>
#include <algorithm>
#include <string_view>

namespace tlsfp {

// Lightweight byte reader with bounds checking
struct ByteReader {
    const uint8_t *data;
    size_t len;
    size_t offset{0};

    inline bool has_bytes(size_t n) const noexcept {
        return (n <= len && offset <= len - n);
    }

    inline uint8_t read_u8() noexcept {
        return data[offset++];
    }

    inline uint16_t read_u16() noexcept {
        uint16_t val = (static_cast<uint16_t>(data[offset]) << 8) | data[offset + 1];
        offset += 2;
        return val;
    }

    inline uint32_t read_u24() noexcept {
        uint32_t val = (static_cast<uint32_t>(data[offset]) << 16) |
                       (static_cast<uint32_t>(data[offset + 1]) << 8)  |
                       data[offset + 2];
        offset += 3;
        return val;
    }

    inline bool skip(size_t n) noexcept {
        if (!has_bytes(n)) return false;
        offset += n;
        return true;
    }
};

// Evaluates if a 2-byte ALPN string is an RFC 8701 GREASE value
inline bool is_alpn_grease(std::string_view alpn) noexcept {
    if (alpn.size() != 2) return false;
    uint8_t b1 = static_cast<uint8_t>(alpn[0]);
    uint8_t b2 = static_cast<uint8_t>(alpn[1]);
    return (b1 == b2) && ((b1 & 0x0F) == 0x0A);
}

// Parse the extensions vector from a ClientHello message
static void parse_extensions(ByteReader &reader, size_t exts_len, ClientHelloData &out) {
    const size_t end_offset = reader.offset + exts_len;

    while (reader.offset + 4 <= end_offset) {
        uint16_t ext_type = reader.read_u16();
        uint16_t ext_len = reader.read_u16();

        // Prevent extension length from overrunning the extension block
        if (reader.offset + ext_len > end_offset) {
            break;
        }

        // Filter GREASE extension types (RFC 8701)
        if (!is_grease(ext_type)) {
            out.extensions.push_back(ext_type);
        }

        const size_t ext_start = reader.offset;

        switch (ext_type) {
            // Extension 0x0000: Server Name Indication (SNI)
            case 0x0000: {
                if (ext_len >= 5) {
                    ByteReader sni_reader{reader.data, ext_start + ext_len, ext_start};
                    uint16_t list_len = sni_reader.read_u16();
                    if (list_len + 2 <= ext_len && sni_reader.has_bytes(3)) {
                        uint8_t name_type = sni_reader.read_u8();
                        uint16_t name_len = sni_reader.read_u16();
                        // 0 = host_name per RFC 6066
                        if (name_type == 0 && name_len <= list_len - 3 && sni_reader.has_bytes(name_len)) {
                            out.has_sni = true;
                            // Zerocopy std::string_view eliminates heap allocation
                            out.sni = std::string_view(reinterpret_cast<const char*>(sni_reader.data + sni_reader.offset), name_len);
                        }
                    }
                }
                break;
            }

            // Extension 0x000a: Supported Groups / Elliptic Curves
            case 0x000a: {
                if (ext_len >= 2) {
                    ByteReader ec_reader{reader.data, ext_start + ext_len, ext_start};
                    uint16_t curves_len = ec_reader.read_u16();
                    // Enforce even-byte alignment and bounds check
                    size_t valid_curves_len = std::min<size_t>(curves_len, ext_len - 2);
                    size_t ec_end = ec_reader.offset + (valid_curves_len & ~1ULL);

                    while (ec_reader.offset + 2 <= ec_end) {
                        uint16_t group = ec_reader.read_u16();
                        if (!is_grease(group)) {
                            out.supported_groups.push_back(group);
                        }
                    }
                }
                break;
            }

            // Extension 0x000b: EC Point Formats
            case 0x000b: {
                if (ext_len >= 1) {
                    ByteReader pt_reader{reader.data, ext_start + ext_len, ext_start};
                    uint8_t formats_len = pt_reader.read_u8();
                    size_t pt_end = pt_reader.offset + std::min<size_t>(formats_len, ext_len - 1);
                    while (pt_reader.offset < pt_end) {
                        out.ec_point_formats.push_back(pt_reader.read_u8());
                    }
                }
                break;
            }

            // Extension 0x000d: Signature Algorithms (Required for JA4_c)
            case 0x000d: {
                if (ext_len >= 2) {
                    ByteReader sig_reader{reader.data, ext_start + ext_len, ext_start};
                    uint16_t sig_len = sig_reader.read_u16();
                    size_t valid_sig_len = std::min<size_t>(sig_len, ext_len - 2);
                    size_t sig_end = sig_reader.offset + (valid_sig_len & ~1ULL);

                    while (sig_reader.offset + 2 <= sig_end) {
                        uint16_t sig_algo = sig_reader.read_u16();
                        if (!is_grease(sig_algo)) {
                            out.signature_algorithms.push_back(sig_algo);
                        }
                    }
                }
                break;
            }

            // Extension 0x0010: ALPN (Required for JA4_a)
            case 0x0010: {
                if (ext_len >= 2) {
                    ByteReader alpn_reader{reader.data, ext_start + ext_len, ext_start};
                    uint16_t list_len = alpn_reader.read_u16();
                    if (list_len + 2 <= ext_len) {
                        const size_t list_end = alpn_reader.offset + list_len;
                        while (alpn_reader.offset < list_end && alpn_reader.has_bytes(1)) {
                            uint8_t proto_len = alpn_reader.read_u8();
                            if (proto_len == 0 || !alpn_reader.has_bytes(proto_len)) break;

                            std::string_view cand(
                                reinterpret_cast<const char*>(alpn_reader.data + alpn_reader.offset), 
                                proto_len
                            );
                            alpn_reader.skip(proto_len);

                            // Skip GREASE ALPN values to prevent JA4 fingerprint drift
                            if (!is_alpn_grease(cand)) {
                                out.first_alpn = cand;
                                break;
                            }
                        }
                    }
                }
                break;
            }

            // Extension 0x002b: Supported Versions (TLS 1.3)
            case 0x002b: {
                if (ext_len >= 1) {
                    ByteReader ver_reader{reader.data, ext_start + ext_len, ext_start};
                    uint8_t versions_len = ver_reader.read_u8();
                    size_t valid_ver_len = std::min<size_t>(versions_len, ext_len - 1);
                    size_t ver_end = ver_reader.offset + (valid_ver_len & ~1ULL);

                    while (ver_reader.offset + 2 <= ver_end) {
                        uint16_t ver = ver_reader.read_u16();
                        if (!is_grease(ver)) {
                            out.supported_versions.push_back(ver);
                        }
                    }
                }
                break;
            }

            default:
                break;
        }

        reader.offset = ext_start + ext_len; // Deterministically advance to next block
    }
}

bool parse_client_hello(const uint8_t *payload, size_t len, ClientHelloData &out) {
    ByteReader reader{payload, len, 0};

    // Record Header (5 bytes)
    if (!reader.has_bytes(5)) return false;
    if (reader.read_u8() != 0x16) return false; // Handshake record type

    reader.skip(2); // Skip legacy record version (0x0301 / 0x0303)
    uint16_t record_len = reader.read_u16();
    if (!reader.has_bytes(record_len)) return false;

    // Handshake Header (4 bytes)
    if (!reader.has_bytes(4)) return false;
    if (reader.read_u8() != 0x01) return false; // ClientHello handshake type

    uint32_t handshake_len = reader.read_u24();
    // Validate handshake framing against record boundary
    if (handshake_len + 4 > record_len || !reader.has_bytes(handshake_len)) {
        return false;
    }
    reader.len = reader.offset + handshake_len;
    // Client Version
    if (!reader.has_bytes(2)) return false;
    out.client_version = reader.read_u16();

    // Skip Client Random (32 bytes)
    if (!reader.skip(32)) return false;

    // Skip Session ID
    if (!reader.has_bytes(1)) return false;
    uint8_t session_id_len = reader.read_u8();
    if (!reader.skip(session_id_len)) return false;

    // Cipher Suites (with GREASE filtering)
    if (!reader.has_bytes(2)) return false;
    uint16_t cipher_suites_len = reader.read_u16();
    if (!reader.has_bytes(cipher_suites_len) || (cipher_suites_len % 2 != 0)) return false;

    const size_t cipher_end = reader.offset + cipher_suites_len;
    while (reader.offset < cipher_end) {
        uint16_t cs = reader.read_u16();
        if (!is_grease(cs)) {
            out.cipher_suites.push_back(cs);
        }
    }

    // Compression Methods
    if (!reader.has_bytes(1)) return false;
    uint8_t compression_len = reader.read_u8();
    if (!reader.skip(compression_len)) return false;

    // Extensions Vector
    if (reader.has_bytes(2)) {
        uint16_t extensions_len = reader.read_u16();
        if (!reader.has_bytes(extensions_len)) {
            return false; // Truncated extensions must fail parsing
        }
        parse_extensions(reader, extensions_len, out);
    }

    return true;
}

bool parse_server_hello(const uint8_t *payload, size_t len, ServerHelloData &out) {
    ByteReader reader{payload, len, 0};

    // Record Header
    if (!reader.has_bytes(5)) return false;
    if (reader.read_u8() != 0x16) return false;

    reader.skip(2); // Skip record version
    uint16_t record_len = reader.read_u16();
    if (!reader.has_bytes(record_len)) return false;

    // Handshake Header
    if (!reader.has_bytes(4)) return false;
    if (reader.read_u8() != 0x02) return false; // ServerHello handshake type

    uint32_t handshake_len = reader.read_u24();
    if (handshake_len + 4 > record_len || !reader.has_bytes(handshake_len)) {
        return false;
    }
    reader.len = reader.offset + handshake_len;
    // Server Version (Default for TLS 1.2 and earlier)
    if (!reader.has_bytes(2)) return false;
    out.server_version = reader.read_u16();
    out.selected_version = out.server_version; // Ground 1: Safe fallback initialization

    // Skip Server Random (32 bytes)
    if (!reader.skip(32)) return false;

    // Skip Session ID
    if (!reader.has_bytes(1)) return false;
    uint8_t session_id_len = reader.read_u8();
    if (!reader.skip(session_id_len)) return false;

    // Selected Cipher Suite
    if (!reader.has_bytes(2)) return false;
    out.selected_cipher = reader.read_u16();

    // Skip Compression Method (1 byte)
    if (!reader.skip(1)) return false;

    // Extensions Vector
    if (reader.has_bytes(2)) {
        uint16_t exts_len = reader.read_u16();
        if (!reader.has_bytes(exts_len)) {
            return false; // Bound validation against buffer over-read
        }
        const size_t end_offset = reader.offset + exts_len;

        while (reader.offset + 4 <= end_offset && reader.has_bytes(4)) {
            uint16_t ext_type = reader.read_u16();
            uint16_t ext_len = reader.read_u16();
            
            if (reader.offset + ext_len > end_offset) {
                break;
            }

            if (!is_grease(ext_type)) {
                out.extensions.push_back(ext_type);
            }

            const size_t ext_start = reader.offset;

            // Extension 0x002b: Supported Versions (TLS 1.3 negotiated version)
            if (ext_type == 0x002b && ext_len == 2 && reader.has_bytes(2)) {
                uint16_t ver = (static_cast<uint16_t>(reader.data[reader.offset]) << 8) |
                                reader.data[reader.offset + 1];
                if (!is_grease(ver)) {
                    out.selected_version = ver;
                }
            }

            if (ext_type == 0x0010 && ext_len >= 2) {
                ByteReader alpn_reader{reader.data, ext_start + ext_len, ext_start};
                uint16_t list_len = alpn_reader.read_u16();
                if (list_len + 2 <= ext_len && alpn_reader.has_bytes(1)) {
                    uint8_t proto_len = alpn_reader.read_u8();
                    if (proto_len > 0 && alpn_reader.has_bytes(proto_len)) {
                        out.first_alpn = std::string_view(
                            reinterpret_cast<const char*>(alpn_reader.data + alpn_reader.offset), 
                            proto_len
                        );
                    }
                }
            }

            reader.offset = ext_start + ext_len;
        }
    }

    return true;
}

} // namespace tlsfp

File: code/python/src/parser.py
from dataclasses import dataclass
import struct

@dataclass(frozen=True)
class ClientHelloFields:
    """All fields extracted from a TLS ClientHello, in original wire order."""
    tls_version: int                     # Handshake body version (e.g. 0x0303 = 771)
    cipher_suites: tuple[int, ...]       # All cipher suite values (pre-GREASE-filter)
    extensions: tuple[int, ...]          # Extension type codes, wire order
    elliptic_curves: tuple[int, ...]     # From supported_groups (ext 0x000a)
    ec_point_formats: tuple[int, ...]    # From ec_point_formats (ext 0x000b)
    server_name: str | None              # SNI hostname, if present
    # Fields for JA4 (future):
    alpn: tuple[str, ...] | None = None
    signature_algorithms: tuple[int, ...] | None = None
    supported_versions: tuple[int, ...] | None = None

@dataclass(frozen=True)
class ServerHelloFields:
    """All fields extracted from a TLS ServerHello."""
    tls_version: int                     # Handshake body version
    cipher_suite: int                    # Single selected cipher
    extensions: tuple[int, ...]          # Extension type codes, wire order
    # Field for JA4S (future):
    supported_version: int | None = None


def parse_tls_record(data: bytes, offset: int = 0) -> tuple[int, int, bytes, int]:
    """
    Parse a TLS record header from `data` at `offset`.
    Returns (content_type, version, fragment_bytes, next_offset).

    Every TLS message is wrapped in a Record layer (5 bytes):
    - Byte 0: Content Type (22 means Handshake, 23 is Application Data, etc.)
    - Bytes 1-2: Legacy Record Version (e.g., 0x0301 for TLS 1.0). JA3 explicitly ignores this version.
    - Bytes 3-4: Fragment Length.
    """
    if len(data) - offset < 5:
        raise ValueError("Insufficient data for TLS record header")
    
    content_type = data[offset]
    version = struct.unpack('!H', data[offset+1:offset+3])[0]
    fragment_length = struct.unpack('!H', data[offset+3:offset+5])[0]
    
    next_offset = offset + 5 + fragment_length
    if len(data) < next_offset:
        raise ValueError("Truncated TLS record fragment")
        
    fragment = data[offset+5:next_offset]
    return content_type, version, fragment, next_offset


def parse_handshake_header(data: bytes, offset: int = 0) -> tuple[int, int, int]:
    """
    Parse a Handshake header from `data` at `offset`.
    Returns (msg_type, length, next_offset).

    Every Handshake message is wrapped in a Handshake layer (4 bytes):
    - Byte 0: Message Type (1 = ClientHello, 2 = ServerHello)
    - Bytes 1-3: Message Length (3 bytes)
    """
    if len(data) - offset < 4:
        raise ValueError("Insufficient data for Handshake header")
        
    msg_type = data[offset]
    # Length is 3 bytes, unpack by padding it to 4 bytes
    length_bytes = b'\x00' + data[offset+1:offset+4]
    length = struct.unpack('!I', length_bytes)[0]
    
    return msg_type, length, offset + 4


def parse_client_hello(data: bytes) -> ClientHelloFields:
    """
    Parse the ClientHello handshake body (after the 4-byte handshake header).

    This is where the JA3 components are extracted. TLS is essentially a massive chain of Type-Length-Value (TLV) structures.

    1. Version (2 bytes): The true TLS handshake version (e.g., 0x0303 for TLS 1.2). This is the first field in a JA3 hash.
    2. Random (32 bytes): We skip this by advancing our offset pointer by 32.
    3. Session ID (Variable): We read 1 byte for the length (N), then skip the next N bytes.
    4. Cipher Suites (Variable): We read 2 bytes for the total length of the cipher suites list (M), then loop M/2 times, unpacking 2-byte chunks (!H). This forms the second field of JA3.
    5. Compression Methods (Variable): Read 1 byte for length, skip that many bytes.
    6. Extensions (Variable): We read 2 bytes for the total extensions length, then enter a while loop.
        - Every extension has a 2-byte Type and 2-byte Length.
        - We record the Extension Type (the third field of JA3).
        - If the Type is 0x0000 (SNI), we parse the hostname to print it in the CLI.
        - If the Type is 0x000a (Supported Groups/Curves), we parse the 2-byte curve IDs (the fourth field of JA3).
        - If the Type is 0x000b (EC Point Formats), we parse the 1-byte formats (the fifth field of JA3).
    """
    
    # Minimum size: version (2) + random (32) + session_id_len (1) = 35 bytes
    if len(data) < 35:
        raise ValueError("ClientHello too short")
        
    tls_version = struct.unpack('!H', data[0:2])[0]
    
    offset = 34 # Skip version(2) and random(32)
    session_id_len = data[offset]
    offset += 1 + session_id_len
    
    if len(data) - offset < 2:
        raise ValueError("Truncated at cipher suites length")
        
    cipher_suites_len = struct.unpack('!H', data[offset:offset+2])[0]
    offset += 2

    if cipher_suites_len % 2 != 0:
        raise ValueError("Invalid cipher suites length")
    
    if len(data) - offset < cipher_suites_len:
        raise ValueError("Truncated cipher suites")
        
    cipher_suites = []
    for i in range(0, cipher_suites_len, 2):
        cipher_suite = struct.unpack('!H', data[offset+i:offset+i+2])[0]
        cipher_suites.append(cipher_suite)
    offset += cipher_suites_len
    
    if len(data) - offset < 1:
        raise ValueError("Truncated at compression methods length")
        
    comp_methods_len = data[offset]
    offset += 1 + comp_methods_len
    if offset > len(data):
        raise ValueError("Truncated compression methods")
    extensions = []
    elliptic_curves = []
    ec_point_formats = []
    server_name = None
    alpn = []
    signature_algorithms = []
    supported_versions = []
    
    # If there is data left, parse extensions
    if offset < len(data):
        if len(data) - offset < 2:
            raise ValueError("Truncated at extensions total length")
            
        ext_total_len = struct.unpack('!H', data[offset:offset+2])[0]
        offset += 2

        if offset + ext_total_len > len(data):
            raise ValueError("Truncated extensions data")

        end_offset = offset + ext_total_len
        while offset + 4 <= end_offset:
            ext_type = struct.unpack('!H', data[offset:offset+2])[0]
            ext_len = struct.unpack('!H', data[offset+2:offset+4])[0]
            offset += 4
            
            if offset + ext_len > end_offset:
                break # Truncated extension data, stop parsing
                
            extensions.append(ext_type)
            ext_data = data[offset:offset+ext_len]
            
            if ext_type == 0x0000: # SNI (0)
                if len(ext_data) >= 5: # list len (2), type (1), name len (2)
                    name_type = ext_data[2]
                    if name_type == 0: # Hostname
                        name_len = struct.unpack('!H', ext_data[3:5])[0]
                        if len(ext_data) >= 5 + name_len:
                            server_name = ext_data[5:5+name_len].decode('utf-8', errors='ignore')
                            
            elif ext_type == 0x000a: # supported_groups (10)
                if len(ext_data) >= 2:
                    grp_list_len = struct.unpack('!H', ext_data[0:2])[0]
                    for i in range(0, grp_list_len, 2):
                        if 2+i+2 <= len(ext_data):
                            curve = struct.unpack('!H', ext_data[2+i:4+i])[0]
                            elliptic_curves.append(curve)
                            
            elif ext_type == 0x000b: # ec_point_formats (11)
                if len(ext_data) >= 1:
                    fmt_list_len = ext_data[0]
                    for i in range(fmt_list_len):
                        if 1+i < len(ext_data):
                            ec_point_formats.append(ext_data[1+i])
                            
            elif ext_type == 0x0010: # ALPN (16)
                if len(ext_data) >= 2:
                    alpn_len = struct.unpack('!H', ext_data[0:2])[0]
                    end = min(len(ext_data), 2 + alpn_len)
                    p = 2
                    while p < end:
                        s_len = ext_data[p]
                        p += 1
                        if p + s_len <= end:
                            alpn_str = ext_data[p:p+s_len].decode('utf-8', errors='ignore')
                            alpn.append(alpn_str)
                        p += s_len
                        
            elif ext_type == 0x000d: # signature_algorithms (13)
                if len(ext_data) >= 2:
                    sig_len = struct.unpack('!H', ext_data[0:2])[0]
                    for i in range(0, sig_len, 2):
                        if 2+i+2 <= len(ext_data):
                            sig_alg = struct.unpack('!H', ext_data[2+i:4+i])[0]
                            signature_algorithms.append(sig_alg)
                            
            elif ext_type == 0x002b: # supported_versions (43)
                if len(ext_data) >= 1:
                    v_len = ext_data[0]
                    for i in range(0, v_len, 2):
                        if 1+i+2 <= len(ext_data):
                            sv = struct.unpack('!H', ext_data[1+i:3+i])[0]
                            supported_versions.append(sv)
                            
            offset += ext_len

    return ClientHelloFields(
        tls_version=tls_version,
        cipher_suites=tuple(cipher_suites),
        extensions=tuple(extensions),
        elliptic_curves=tuple(elliptic_curves),
        ec_point_formats=tuple(ec_point_formats),
        server_name=server_name,
        alpn=tuple(alpn) if alpn else None,
        signature_algorithms=tuple(signature_algorithms) if signature_algorithms else None,
        supported_versions=tuple(supported_versions) if supported_versions else None
    )


def parse_server_hello(data: bytes) -> ServerHelloFields:
    """
    Parse the ServerHello handshake body (after the 4-byte handshake header).
    """
    if len(data) < 38:
        raise ValueError("ServerHello too short")
        
    tls_version = struct.unpack('!H', data[0:2])[0]
    
    offset = 34 # Skip version(2) and random(32)
    session_id_len = data[offset]
    offset += 1 + session_id_len
    
    if len(data) - offset < 2:
        raise ValueError("Truncated at cipher suite")
        
    cipher_suite = struct.unpack('!H', data[offset:offset+2])[0]
    offset += 2
    
    if len(data) - offset < 1:
        raise ValueError("Truncated at compression method")
    offset += 1
    
    extensions = []
    supported_version = None
    
    # If there is data left, parse extensions
    if offset < len(data):
        if len(data) - offset < 2:
            raise ValueError("Truncated at extensions total length")
            
        ext_total_len = struct.unpack('!H', data[offset:offset+2])[0]
        offset += 2
        
        end_offset = min(offset + ext_total_len, len(data))
            
        while offset + 4 <= end_offset:
            ext_type = struct.unpack('!H', data[offset:offset+2])[0]
            ext_len = struct.unpack('!H', data[offset+2:offset+4])[0]
            offset += 4
            
            if offset + ext_len > end_offset:
                break
                
            extensions.append(ext_type)
            ext_data = data[offset:offset+ext_len]
            
            if ext_type == 0x002b: # supported_versions (43) in ServerHello is just 2 bytes
                if len(ext_data) >= 2:
                    supported_version = struct.unpack('!H', ext_data[0:2])[0]
                    
            offset += ext_len

    return ServerHelloFields(
        tls_version=tls_version,
        cipher_suite=cipher_suite,
        extensions=tuple(extensions),
        supported_version=supported_version
    )
